[IA][RISCV] Support VP loads/stores in InterleavedAccessPass

mshockwave opened the pull request:
Teach InterleavedAccessPass to recognize the following patterns:
  - vp.store an interleaved scalable vector
  - Deinterleaving a scalable vector loaded from vp.load

Upon recognizing these patterns, IA will collect the interleaved / deinterleaved operands and delegate them over to their respective newly-added TLI hooks.

For RISC-V, these patterns are lowered into segmented loads/stores (except when we're interleaving constant splats, in which case a unit-strde store will be generated)

Right now we only recognized power-of-two (de)interleave cases, in which (de)interleave4/8 are synthesized from a tree of (de)interleave2.

mshockwave commented on llvm/test/CodeGen/RISCV/rvv/scalable-vectors-interleaved-access.ll:
```diff
@@ -0,0 +1,512 @@
+; RV32:       # %bb.0:
+; RV32-NEXT:    addi sp, sp, -16
+; RV32-NEXT:    .cfi_def_cfa_offset 16
+; RV32-NEXT:    li a1, 777
+; RV32-NEXT:    li a2, 666
+; RV32-NEXT:    sw a2, 8(sp)
+; RV32-NEXT:    sw a1, 12(sp)
+; RV32-NEXT:    addi a1, sp, 8
+; RV32-NEXT:    vsetvli a2, zero, e64, m8, ta, ma
+; RV32-NEXT:    vlse64.v v8, (a1), zero
```
Will it be better to use vmerge here?

topperc replied:
I'm not sure how vmerge helps. Can you elaborate?

mshockwave replied:
```
li a1, 666
vsetvli a2, zero, e32, m8
vmv.v.x v8, a1
vmerge.vim v8, v8, 777, <mask of 01010101....>
```

topperc commented on llvm/test/CodeGen/RISCV/rvv/scalable-vectors-interleaved-access.ll:
```diff
@@ -0,0 +1,512 @@
+  ret void
+}
+
+; Expecting uni-strde store here rather than segmented store.
```
stride*

topperc replied:
unit*

mshockwave replied:
Typing is hard :-P . It's fixed now

topperc commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -248,6 +249,186 @@ static bool isReInterleaveMask(ShuffleVectorInst *SVI, unsigned &Factor,
+
+    unsigned VisitedIdx = 0;
+    for (User *Usr : Current->users()) {
+      // We're playing safe here and matches only the expression
```
matches -> matching

mshockwave replied:
Fixed.

topperc commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -248,6 +249,186 @@ static bool isReInterleaveMask(ShuffleVectorInst *SVI, unsigned &Factor,
+      // on the same index. So we XOR (index + 1) onto VisitedIdx
+      // such that if there is any duplication, VisitedIdx will be
+      // zero.
+      VisitedIdx ^= Indices[0] + 1;
```
Should `Indices[0] + 1` be `1 << Indices[0]`? Seems like you're trying to treat VisitedIdx as a bit vector but then taking shortcuts because there are only 2 bits?

mshockwave replied:
Yes, it just so happens that `1 << Indices[0]` has the same effect as `Indices[0] + 1` but I am trying to do a bit vector here. It's fixed now.

topperc commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -248,6 +249,186 @@ static bool isReInterleaveMask(ShuffleVectorInst *SVI, unsigned &Factor,
+    unsigned VisitedIdx = 0;
+    for (User *Usr : Current->users()) {
+      // We're playing safe here and matches only the expression
+      // consisting of a perfectly balanced binary tree in which all
```
Where do verify it is balanced?

mshockwave replied:
It's done. I check the leaf types.

hassnaaHamdi commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -248,6 +249,196 @@ static bool isReInterleaveMask(ShuffleVectorInst *SVI, unsigned &Factor,
+//  to reorder them by interleaving these values.
+static void interleaveLeafValues(SmallVectorImpl<Value *> &Leaves) {
+  unsigned Factor = Leaves.size();
+  assert(isPowerOf2_32(Factor) && Factor <= 8 && Factor > 1);
```
Hi @mshockwave 
I have done a simpler approach to support interleaving of factor 4 by using pattern match.
In my opinion, I think it will be better to do either the simple approach to support factor 8, or your current approach but with support to all power_of_2 factors.
I think the BFS approach is better for all power_of_2 factors, while the simple approach (pattern match) is better for only limited factors like 4 or 8.
And I see that you already did the heavy work, so making it generic for all power_of_2 factors would be better.
That is my opinion :'D

mshockwave replied:
> or your current approach but with support to all power_of_2 factors.
> so making it generic for all power_of_2 factors would be better.

Oh yeah absolutely, I added `Factor <= 8` here simply because RISC-V (and I believe AArch64?) support at most 8. But that said, I'm happy to relax this restriction, in which case I literally only need to remove this condition and it should support all power-of-two out of the box.

lukel97 commented:
Thanks for working on this, this will be important for EVL tail folding.

I noticed though that we don't support factor > 2 [de]interleaves for non-VP scalable loads and stores on RISC-V yet.

It seems strange that we would support it for VP ops but not for regular ops, could this PR be reworked to maybe add support for the latter first? I.e. consolidating the target-independent BFS here + the AArch64 TLI pattern matching.

lukel97 commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -248,6 +249,196 @@ static bool isReInterleaveMask(ShuffleVectorInst *SVI, unsigned &Factor,
+      // on the same index. So we XOR (1 << index) onto VisitedIdx
+      // such that if there is any duplication, VisitedIdx will be
+      // zero.
+      VisitedIdx ^= (1 << Indices[0]);
+      if (!VisitedIdx)
+        return 0;
```
Given that there's only two users, would it be easier to identify the LHS + RHS first? Then we can be explicitly do the BFS from left to right and not have to swap

```c++
ExtractValueInst *LHS, RHS;
for (User *Usr : Current->users()) {
  if (Indices[0] == 0 && !LHS)
    LHS = Usr;
  else if (Indices[0] == 1 && !RHS)
    RHS = Usr;
  else
    return 0;
}

for (ExtractValueInst EV : {LHS, RHS}) {
  // continue the traversal
  if (match(...)) {
    Queue.push_back(EVUsr);
    continue;
  }
  // check if balanced...
  Results.push_back(EV);
}
```

mshockwave replied:
Good point. It's fixed now.

lukel97 commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -248,6 +249,196 @@ static bool isReInterleaveMask(ShuffleVectorInst *SVI, unsigned &Factor,
+        return Operands.front();
+    }
+  }
+  if (match(WideMask, m_AllOnes()))
+    return nullptr;
```
I think if you pass on the all ones mask to RISC-V's TTI, RISCVVectorPeephole will convert it to its unmasked equivalent anyway. Or at least it should. So you might be able to get away with only emitting the masked intrinsic.

mshockwave replied:
> RISCVVectorPeephole will convert it to its unmasked equivalent anyway. Or at least it should. 

I think `RISCVDAGToDAGISel::doPeepholeMaskedRVV` will kick in before RISCVVectorPeephole. But in any case, #123106 adds the missing RISCVMaskedPseudo that used to prevent that from happening.

mshockwave replied:
> So you might be able to get away with only emitting the masked intrinsic.

It's done now

lukel97 replied:
Ah I forgot we don't handle loads and stores yet, thanks for submitting that. I don't see any reason why we can't handle them too.

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -22190,6 +22190,275 @@ bool RISCVTargetLowering::lowerInterleaveIntrinsicToStore(
+
+  // Tail-policy
+  if (Mask)
+    Operands.push_back(ConstantInt::get(XLenTy, 1));
```
```suggestion
    Operands.push_back(ConstantInt::get(XLenTy, RISCVII::TAIL_AGNOSTIC));
```

mshockwave replied:
Fixed.

mshockwave commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -248,6 +249,196 @@ static bool isReInterleaveMask(ShuffleVectorInst *SVI, unsigned &Factor,
+    //  A B C D E F G H
+    for (unsigned SrcIdx : {0, 4, 2, 6, 1, 5, 3, 7})
+      Buffer.push_back(Leaves[SrcIdx]);
+  }
```
Yes, it has been generalized now.

mshockwave commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -248,6 +249,196 @@ static bool isReInterleaveMask(ShuffleVectorInst *SVI, unsigned &Factor,
+//  getVectorInterleaveFactor / getVectorDeinterleaveFactor. But TLI
+//  hooks (e.g. lowerInterleavedScalableLoad) expect ABCD, so we need
+//  to reorder them by interleaving these values.
+static void interleaveLeafValues(SmallVectorImpl<Value *> &Leaves) {
```
Leaves (or SubLeaves as I just pushed another batch of changes) is both the input and the output container.

topperc replied:
Oops. Sorry about that. I forget that std::copy has the destination at the end.

mshockwave commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -248,6 +249,196 @@ static bool isReInterleaveMask(ShuffleVectorInst *SVI, unsigned &Factor,
+//  to reorder them by interleaving these values.
+static void interleaveLeafValues(SmallVectorImpl<Value *> &Leaves) {
+  unsigned Factor = Leaves.size();
+  assert(isPowerOf2_32(Factor) && Factor <= 8 && Factor > 1);
```
I just generalized interleaveLeafValues so it now handles all power-of-two cases.

mshockwave commented:
I just stack this PR on top of #123106 so that we now only need to generate masked version of segmented load/store.

mshockwave commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -248,6 +249,196 @@ static bool isReInterleaveMask(ShuffleVectorInst *SVI, unsigned &Factor,
+      // into Queue/Results if necessary (This is also one of the main
+      // reasons using BFS instead of DFS here, btw).
+
+      // When VisitedIdx equals to 0b11, we're the last visted ExtractValue.
```
I'm no longer using VisitedIdx in this version.

mshockwave commented:
> It seems strange that we would support it for VP ops but not for regular ops, could this PR be reworked to maybe add support for the latter first? I.e. consolidating the target-independent BFS here + the AArch64 TLI pattern matching.

I'm going to circle back to this later. I have addressed rest of your comments though

mshockwave committed:
[IA][RISCV] Support VP intrinsics in InterleavedAccessPass

Teach InterleavedAccessPass to recognize the following patterns:

  - vp.store an interleaved scalable vector
  - Deinterleaving a scalable vector loaded from vp.load

Upon recognizing these patterns, IA will collect the interleaved /
deinterleaved operands and delegate them over to their respective
newly-added TLI hooks.

For RISC-V, these patterns are lowered into segmented loads/stores
(except when we're interleaving constant splats, in which case a
unit-strde store will be generated)

Right now we only recognized power-of-two (de)interleave cases, in which
(de)interleave4/8 are synthesized from a tree of (de)interleave2.

Co-authored-by: Nikolay Panchenko <nicholas.panchenko@gmail.com>

mshockwave commented:
> Thanks for working on this, this will be important for EVL tail folding.
> 
> I noticed though that we don't support factor > 2 [de]interleaves for non-VP scalable loads and stores on RISC-V yet.
> 
> It seems strange that we would support it for VP ops but not for regular ops, could this PR be reworked to maybe add support for the latter first? I.e. consolidating the target-independent BFS here + the AArch64 TLI pattern matching.

I have split the generalized power-of-two (de)interleave support for non-vp load/store into #123863 . I have also updated this PR to reflect the change.

Together, now both AArch64 and RISC-V share the same factor calculation logics that is capable of analyzing any power-of-two (de)interleave intrinsics.

lukel97 commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -630,11 +630,34 @@ getVectorDeinterleaveFactor(IntrinsicInst *II,
+    }
+  }
+  if (match(WideMask, m_AllOnes()))
+    return nullptr;
```
If you return WideMask here then you can remove the optional and return nullptr on failure. Then you also don't need to handle the null Mask value in lowerInterleavedScalableLoad/Store

mshockwave replied:
Good catch. It's fixed now

lukel97 commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -643,43 +666,92 @@ bool InterleavedAccessImpl::lowerDeinterleaveIntrinsic(
+
+    // Since lowerInterleaveLoad expects Shuffles and LoadInst, use special
+    // TLI function to emit target-specific interleaved instruction.
+    if (!TLI->lowerInterleavedScalableLoad(VPLoad, *Mask, DI,
+                                           DeinterleaveValues))
```
Is there anything preventing this from being called with a fixed length vector? Or why do we need to restrict it to scalable vectors in the first place?

It looks like RISCVTargetLowering::lowerDeinterleaveIntrinsicToLoad already handles fixed length vectors.

I don't think we need to handle fixed length vectors in this PR necessarily, we could just check for them in the TII hook and return false.

topperc replied:
The vectorizer shouldn't generate deinterleave/interleave intrinsics for fixed vectors, but maybe we should guard against handwritten IR?

mshockwave replied:
Done. Now RISCV's TLI implementation will check if it's scalable vectors or now.

lukel97 commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -3152,6 +3153,34 @@ class TargetLoweringBase {
+  /// \p DeinterleaveIntrin is vector.deinterleave intrinsic
+  /// \p DeinterleaveRes is a list of deinterleaved results.
+  virtual bool
+  lowerInterleavedScalableLoad(VPIntrinsic *Load, Value *Mask,
```
If this doesn't have to be strictly restricted to scalable vectors, would it make sense to rename this something like `lowerDeinterleaveIntrinsicToVPLoad` to more closely match the `lowerDeinterleaveIntrinsicToLoad` hook?

mshockwave replied:
I also updated the list of arguments taken by lowerDeinterleaveIntrinsicToVPLoad / lowerInterleaveIntrinsicToVPStore and they no longer take the (de)interleaved intrinsic argument (because we're not really using it)

mshockwave committed:
fixup! Address review comments

  - Change what getMask returns
  - Change the name of TLI hooks
  - Limit the TLI hooks to scalable vectors only for now

mshockwave committed:
fixup! Rename test

And add negative test case for fixed vectors.

topperc commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -630,11 +630,35 @@ getVectorDeinterleaveFactor(IntrinsicInst *II,
+    if (getVectorInterleaveFactor(IMI, Operands, DeadInsts)) {
+      assert(!Operands.empty());
+      if (Operands.size() == Factor &&
+          std::equal(Operands.begin(), Operands.end(), Operands.begin()))
```
I'm not sure this std::equal does what you think it does.

It's going to create 2 iterators that both point to Operands.begin() and then compare them to each other for every element. They're always going to match.

mshockwave replied:
Oops, it's fixed now. I also added a negative test case.

mshockwave committed:
fixup! Address review comments

mshockwave committed:
fixup! fixup! Address review comments

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -22529,6 +22529,261 @@ bool RISCVTargetLowering::lowerInterleaveIntrinsicToStore(
+  IRBuilder<> Builder(Load);
+  Value *WideEVL = Load->getArgOperand(2);
+  auto *XLenTy = Type::getIntNTy(Load->getContext(), Subtarget.getXLen());
+  Value *EVL = Builder.CreateZExtOrTrunc(
+      Builder.CreateUDiv(WideEVL, ConstantInt::get(WideEVL->getType(), Factor)),
+      XLenTy);
```
Is this transform correct if we end up truncating the original EVL? 

E.g. if the original wide vp load had 7 elements, this will transform it into only loading 6 elements, making the last element undefined.

I have a feeling that whenever the loop vectorizer adds EVL support for interleaved accesses the EVL passed to the wide vp.load will be `@llvm.experimental.get.vector.length.i64` multiplied by the factor or something.

Can we get away with checking that the EVL is a multiple of the factor with `computeKnownBits`? So we could only handle test cases for now where the wide vp.load's EVL is a mul of the factor. Or is there extra alignment information about the segment that would allow us to increase the EVL?

lukel97 replied:
For non power of 2s we could explicitly match a mul on the EVL. cc) @Mel-Chen, have you begun work on the LoopVectorizer part of this yet in #123201? Do you know what will be passed to the EVL argument of the wide vp load?

mshockwave replied:
I'd added a really simple check on the EVL using either computeKnownBits or explicit pattern matching on`mul`.

lukel97 commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -630,11 +630,34 @@ getVectorDeinterleaveFactor(IntrinsicInst *II,
+    }
+  }
+  if (match(WideMask, m_AllOnes()))
+    return WideMask;
```
Oh I see it's because after the tuple types in #97992 we don't have another vector to add attach a `LLVMScalarOrSameVectorWidth` type to. I'm not sure if there's an easy fix for this cc) @4vtomat

mshockwave replied:
> Oh I see it's because after the tuple types in #97992 we don't have another vector to add attach a `LLVMScalarOrSameVectorWidth` type to. I'm not sure if there's an easy fix for this cc) @4vtomat

A potential solution would be allowing us to supply a custom code fragment to `LLVMScalarOrSameVectorWidth` to retrieve the vector type we want to match against. For instance:
```
LLVMScalarOrSameVectorWidth<0, [{ return cast<TargetExtType>(Ty)->getTypeParameter(0); }], llvm_i1_ty>;
```
where `Ty` is an implicit `Type *` variable representing the original intrinsic parameter type.

lukel97 replied:
That sounds good. I think that's out of the scope for this PR though, if we can just manually fix WideMask to return the segment type here I think that's fine.

mshockwave replied:
> if we can just manually fix WideMask to return the segment type here I think that's fine.

Yeah, I just added that, along with checking if the (non-all-ones) mask has the desired vector length.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -22529,6 +22529,261 @@ bool RISCVTargetLowering::lowerInterleaveIntrinsicToStore(
+
+  Value *PoisonVal = PoisonValue::get(VecTupTy);
+  SmallVector<Value *> Operands;
+  Operands.append({PoisonVal, Load->getArgOperand(0)});
```
This can append can be combined into the the SmallVector constructor above

mshockwave replied:
Fixed

mshockwave committed:
fixup! Address some of the review comments

mshockwave commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -22529,6 +22529,261 @@ bool RISCVTargetLowering::lowerInterleaveIntrinsicToStore(
+  Operands.push_back(EVL);
+
+  // Tail-policy
+  Operands.push_back(ConstantInt::get(XLenTy, RISCVII::TAIL_AGNOSTIC));
```
Fixed.

mshockwave commented on llvm/test/CodeGen/RISCV/rvv/vp-vector-interleaved-access.ll:
```diff
@@ -0,0 +1,756 @@
+; RV64-NEXT:    vsetvli zero, a1, e32, m1, ta, ma
+; RV64-NEXT:    vlseg2e32.v v8, (a0)
+; RV64-NEXT:    ret
+  %wide.masked.load = call <vscale x 4 x i32> @llvm.vp.load.nxv4i32.p0(ptr %ptr, <vscale x 4 x i1> shufflevector (<vscale x 4 x i1> insertelement (<vscale x 4 x i1> poison, i1 true, i32 0), <vscale x 4 x i1> poison, <vscale x 4 x i32> zeroinitializer), i32 %rvl)
```
Done

mshockwave commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -630,11 +630,34 @@ getVectorDeinterleaveFactor(IntrinsicInst *II,
+///   deinterleaved mask
+/// - if a value within the option is nullptr, the value corresponds to all-true
+///   mask
+/// - return nullopt if mask cannot be deinterleaved
```
Fixed.

mshockwave committed:
fixup! Address review comments

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -22529,6 +22530,282 @@ bool RISCVTargetLowering::lowerInterleaveIntrinsicToStore(
+  if (isPowerOf2_32(N)) {
+    KnownBits KB = llvm::computeKnownBits(V, DL);
+    return KB.countMinTrailingZeros() >= Log2_32(N);
+  } else {
```
No else after return.

mshockwave replied:
Fixed.

mshockwave committed:
fixup! Add more comments

mshockwave committed:
fixup! Address review comments

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -22529,6 +22530,286 @@ bool RISCVTargetLowering::lowerInterleaveIntrinsicToStore(
+                              NumElts * SEW / 8),
+      Factor);
+
+  Value *PoisonVal = PoisonValue::get(VecTupTy);
+  SmallVector<Value *> Operands{PoisonVal, Load->getArgOperand(0)};
```
Nit, move this down beside the rest of the Operands.push_backs?

mshockwave replied:
Done. I use a normal array instead now.

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -22529,6 +22530,286 @@ bool RISCVTargetLowering::lowerInterleaveIntrinsicToStore(
+    Return = Builder.CreateInsertValue(Return, VecExtract, i);
+  }
+
+  for (auto [Idx, DIO] : enumerate(DeInterleaveResults)) {
```
Nit
```suggestion
  for (auto [Idx, DIO] : enumerate(DeinterleaveResults)) {
```

mshockwave replied:
Fixed.

lukel97 commented on llvm/lib/CodeGen/InterleavedAccessPass.cpp:
```diff
@@ -630,11 +630,38 @@ getVectorDeinterleaveFactor(IntrinsicInst *II,
+    // Scale the vector length of all-ones mask.
+    ElementCount OrigEC =
+        cast<VectorType>(WideMask->getType())->getElementCount();
+    if (OrigEC.getKnownMinValue() % Factor)
+      return nullptr;
```
Is this an invariant? Can this be made into an assert?

mshockwave replied:
Yes they're invariant because getVector(De)InterleaveFactor + vp.load/store signature have make sure it'll be a multiple of Factor. It's fixed now.

mshockwave commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -22529,6 +22530,286 @@ bool RISCVTargetLowering::lowerInterleaveIntrinsicToStore(
+/// into
+/// `<vscale x 8 x i32>`. This will resuling a simple unit stride store rather
+/// than a segment store, which is more expensive in this case.
+static Value *foldInterleaved2OfConstSplats(Value *Op0, Value *Op1,
```
> This looks like a useful transform to have, did this show up in any workloads that you were seeing?

Yes it was found by @topperc in 482.sphinx3 a while ago

> 
> I wonder if it's easier to implement this as a peephole on vector.interleave2 in VectorCombine, and then it becomes independent of the store.

I think it's reasonable to move to VectorCombine.

(And I just realized the comments are wrong, it should be "...create a larger splat `<vscale x 8 x i64> <splat of ((777 << 32) | 666)>` first before casting it into `<vscale x 16 x i32>`...")

mshockwave replied:
Okay, I've split out this part into #125144

mshockwave committed:
fixup! Address review comments

mshockwave commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -22529,6 +22530,286 @@ bool RISCVTargetLowering::lowerInterleaveIntrinsicToStore(
+  // FIXME: Should pass alignment attribute from pointer, but vectorizer needs
+  // to emit it first.
+  auto &DL = Load->getModule()->getDataLayout();
+  Align Alignment = Align(DL.getTypeStoreSize(WideVTy->getScalarType()));
```
Good point, it's fixed now.

mshockwave committed:
fixup! Split out the folding rule on interleave2 of two const splats

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -22529,6 +22530,233 @@ bool RISCVTargetLowering::lowerInterleaveIntrinsicToStore(
+    StoredVal = Builder.CreateCall(
+        VecInsertFunc, {StoredVal, InterleaveOperands[i], Builder.getInt32(i)});
+
+  SmallVector<Value *, 5> Operands;
+  Operands.push_back(StoredVal);
+  Operands.push_back(Store->getArgOperand(1));
```
Nit, also make these operands a plain array same as the loads?

mshockwave replied:
Fixed.

lukel97 commented on llvm/test/CodeGen/RISCV/rvv/vp-vector-interleaved-access.ll:
```diff
@@ -0,0 +1,816 @@
+  %interleaved.mask = tail call <vscale x 4 x i1> @llvm.vector.interleave2.nxv4i1(<vscale x 2 x i1> %mask, <vscale x 2 x i1> %mask)
+  %wide.masked.load = tail call <vscale x 4 x i32> @llvm.vp.load.nxv4i32.p0(ptr %ptr, <vscale x 4 x i1> %interleaved.mask, i32 %rvl)
+  %deinterleaved.results = tail call { <vscale x 2 x i32>, <vscale x 2 x i32> } @llvm.vector.deinterleave2.nxv16i32(<vscale x 4 x i32> %wide.masked.load)
+  %t0 = extractvalue { <vscale x 2 x i32>, <vscale x 2 x i32> } %deinterleaved.results, 0
+  %t1 = extractvalue { <vscale x 2 x i32>, <vscale x 2 x i32> } %deinterleaved.results, 1
```
Is the test for this to make sure we don't lower the segmented load if there's another use of the wide load? Do we need to add another identity insertvalue pair for it to be recognised?

lukel97 commented on llvm/test/CodeGen/RISCV/rvv/vp-vector-interleaved-access.ll:
```diff
@@ -0,0 +1,816 @@
+  %t1 = extractvalue { <vscale x 2 x i32>, <vscale x 2 x i32> } %d2, 0
+  %t3 = extractvalue { <vscale x 2 x i32>, <vscale x 2 x i32> } %d2, 1
+
+  %res0 = insertvalue { <vscale x 2 x i32>, <vscale x 2 x i32>, <vscale x 2 x i32>, <vscale x 2 x i32> } undef, <vscale x 2 x i32> %t0, 0
```
Should probably replace undef with poison in this file

mshockwave replied:
Fixed.

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -22529,6 +22530,233 @@ bool RISCVTargetLowering::lowerInterleaveIntrinsicToStore(
+///             @llvm.vector.deinterleave2.nxv64i8(
+///               <vscale x 64 x i8> %l)
+///   %r0 = extractvalue { <vscale x 32 x i8>, <vscale x 32 x i8> } %dl, 0
+///   %r1 = extractvalue { <vscale x 32 x i8>, <vscale x 32 x i8> } %dl, 1
```
Nit, should we add the insertvalues into the example here? And maybe leave a TODO to match factor 2 deinterleaves without them?

mshockwave replied:
> should we add the insertvalues into the example here?

insertvalue is actually not part of the pattern, only extractvalue is. We use insertvalue in codegen tests only to form a returned value from the deinterleaved parts, otherwise the entire function would be DCE (As a counter example, `test/Transforms/InterleavedAccess/RISCV/interleaved-accesses.ll` doesn't use insertvalue).

> And maybe leave a TODO to match factor 2 deinterleaves without them?

Done

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -22529,6 +22530,233 @@ bool RISCVTargetLowering::lowerInterleaveIntrinsicToStore(
+    return false;
+
+  auto *XLenTy = Type::getIntNTy(Store->getContext(), Subtarget.getXLen());
+  Value *EVL = Builder.CreateZExtOrTrunc(
```
Nit, WideEVL should always be i32 and XLenTy should always be >= i32, so I don't think this should ever truncate. Make this just Builder.CreateZExt?

mshockwave replied:
Fixed.

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -22529,6 +22530,233 @@ bool RISCVTargetLowering::lowerInterleaveIntrinsicToStore(
+    return false;
+
+  auto *XLenTy = Type::getIntNTy(Load->getContext(), Subtarget.getXLen());
+  Value *EVL = Builder.CreateZExtOrTrunc(
```
Nit, WideEVL should always be i32 and XLenTy should always be >= i32, so I don't think this should ever truncate. Make this just Builder.CreateZExt?

mshockwave replied:
Fixed.

lukel97 reviewed (approved):
LGTM with nits, thanks

mshockwave committed:
fixup! Address review comments
